### PR TITLE
Add mkST and runST family of functions

### PR DESCRIPTION
**Pull request by mikesol**
**Description of the change**

This PR is a copy-paste of Effect's uncurried logic.

---

**Checklist:**

- [x] Added a file to CHANGELOG.d for this PR (see CHANGELOG.d/README.md)
- [x] Added myself to CONTRIBUTORS.md (if this is my first contribution)
- [ ] Linked any existing issues or proposals that this pull request should close
- [ ] Updated or added relevant documentation
- [x] Added a test for the contribution (if applicable)

### CODE REVIEW

**Comment by rhendric** [comment]
Looks like this is a companion to purescript/purescript-st#52.

A test would be nice, along the lines of previous iteration #3979. Otherwise, no objections here, conditional on purescript/purescript-st#52 actually landing this time.

**Comment by mikesol** [comment]
> A test would be nice, along the lines of previous iteration #3979.

Is there a way to assert that the codegen is correct? Testing the functionality is important too (I'll add that), but I'd also like to assert that the generated code is actually optimized.

**Comment by rhendric** [comment]
Yeah, have a look in tests/purs/optimize.

**Comment by mikesol** [comment]
> Yeah, have a look in tests/purs/optimize.

Awesome, thanks for the tip!

Would it make sense, then, to wait on this PR until https://github.com/purescript/purescript-st/pull/52 lands?

**Comment by rhendric** [comment]
Personally, that's the order in which I think things should happen, contra (!) the previous ruling in purescript/purescript-st#36. Mention me if this causes a deadlock.

**Comment by JordanMartinez** [comment]
The ST PR has been merged. I think tomorrow, the latest package set will also include the new code in `st`, and this should then be unblocked.

**Comment by mikesol** [comment]
I added a test & it passes. I think it's good to go from a functionality standpoint, but lemme know if there's anything I'm missing in terms of style/form.

File: tests/purs/optimize/4386.purs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+mySTFn1 = mkSTFn1 \a -> pure (a + 1)
+
+mySTFn2 :: forall r. STFn2 Int Int r Int
+mySTFn2 = mkSTFn2 \a b -> pure (a + b)
+
+myInt1 :: forall r. ST r Int
+myInt1 = runSTFn1 mySTFn1 0
+
+myInt2 :: forall r. ST r Int
+myInt2 = runSTFn2 mySTFn2 0 1
```
Comment (JordanMartinez) [inline tests/purs/optimize/4386.purs:17]: Should the below code be added here to ensure we get MagicDo inlining as well?

```purs
otherTest :: forall r. ST r Int
otherTest = do
  a <- runSTFn2 mySTFn2 0 1
  b <- runSTFn1 mySTFn1 2
  c <- myInt1
  d <- myInt2
  pure $ a + b + c + d
```
Testing this out locally, I get:
```
var otherTest = function __do() {
    var a = mySTFn2(0, 1);
    var b = mySTFn1(2);
    var c = myInt1();
    var d = myInt2();
    return ((a + b | 0) + c | 0) + d | 0;
};
```

File: CHANGELOG.d/feature_optimize-uncurried-st.md [hunk 2]
```diff
@@ -0,0 +1 @@
+* Adds a compiler optimization for `ST` functions with up to 10 arity, similar to `Effect` optimizations.
```
Comment (rhendric) [inline CHANGELOG.d/feature_optimize-uncurried-st.md:1]: Change ‘Adds’ to ‘Add’, please? The changelog uses the same imperative mood as is recommended for Git commit messages.

File: tests/purs/optimize/4386.purs [hunk 3]
```diff
@@ -15,3 +15,11 @@ myInt1 = runSTFn1 mySTFn1 0
...
 myInt2 :: forall r. ST r Int
 myInt2 = runSTFn2 mySTFn2 0 1
+
+otherTest :: forall r. ST r Int
+otherTest = do
+  a <- runSTFn2 mySTFn2 0 1
+  b <- runSTFn1 mySTFn1 2
+  c <- myInt1
+  d <- myInt2
+  pure $ a + b + c + d
```
Comment (JordanMartinez) [inline tests/purs/optimize/4386.purs:25]: Small nit, but could you add a blank newline to the end of this file?

**Review by JordanMartinez — APPROVED** [review]

File: tests/purs/optimize/4386.purs (hunk 3 above)
Comment (mikesol) [reply tests/purs/optimize/4386.purs]: https://github.com/purescript/purescript/pull/4386/commits/73fac965e9406f7995899c51cdfc3755f1bf5e19

Comment (JordanMartinez) [reply tests/purs/optimize/4386.purs]: Thanks!

**Review by rhendric — APPROVED** [review]

**Pull request merged**